legate/core/runtime.py

```diff
@@ -586,24 +632,26 @@ def detach_external_allocation(
         detach: Union[Detach, IndexDetach],
         defer: bool = False,
         previously_deferred: bool = False,
-    ) -> None:
+    ) -> Union[None, Future]:
         # If the detachment was previously deferred, then we don't
         # need to remove the allocation from the map again.
         if not previously_deferred:
             self._remove_allocation(alloc)
         if defer:
             # If we need to defer this until later do that now
+            # IRINA FIXME should I pass unordered here?
             self._deferred_detachments.append((alloc, detach))
```

Comment:
I think you need to record the `RegionField`'s `self.field` in this scenario, so when the detachment actually happens later in `perform_detachments` you can call `Field.add_detach_future` at that point.

I would pass a `Field` argument to this method, change this line to

```
self._deferred_detachments.append((alloc, detach, field))
```

change the type of `AttachmentManager._deferred_detachments` appropriately, and change `perform_detachments` as follows:

```
    def perform_detachments(self) -> None:
        detachments = self._deferred_detachments
        self._deferred_detachments = list()
        for alloc, detach, field in detachments:
            detach_future = self.detach_external_allocation(
                alloc, detach, defer=False, previously_deferred=True
            )
            if detach_future is not None:
                field.add_detach_future(detach_future)
```